### PR TITLE
gobot formfilling

### PR DESCRIPTION
**Pull request by oserikov**

### CODE REVIEW

File: deeppavlov/dataset_readers/md_yaml_dialogs_reader.py [hunk 1]
```diff
@@ -241,78 +246,96 @@ def _read_story(cls,
...
+                    utters_to_append_batch.append([user_utter])
+
+            except KeyError as e:
+                log.debug(f"INSIDE MLU_MD_DialogsDatasetReader._read_story(): "
+                          f"Skipping story w. line {line} because of no NLU candidates found")
+                nonlocal_curr_story_bad = True
+                utters_to_append_batch = []
+            return utters_to_append_batch
+
+        def process_system_utter(line):
```
Comment (oserikov) [inline deeppavlov/dataset_readers/md_yaml_dialogs_reader.py:271]: todo: docs, typehints

File: deeppavlov/dataset_readers/md_yaml_dialogs_reader.py [hunk 2]
```diff
@@ -241,78 +246,96 @@ def _read_story(cls,
 
         curr_story_title = None
         curr_story_utters_batch = []
-        curr_story_bad = False
-        for line in open(story_fpath):
+        nonlocal_curr_story_bad = False  # can be modified as a nonlocal variable
+
+        def process_user_utter(line):
```
Comment (oserikov) [inline deeppavlov/dataset_readers/md_yaml_dialogs_reader.py:251]: todo: docs, typehints

File: deeppavlov/dataset_readers/md_yaml_dialogs_reader.py [hunk 3]
```diff
@@ -241,78 +246,96 @@ def _read_story(cls,
...
+            nonlocal intent2slots2text, domain_knowledge, curr_story_utters_batch, nonlocal_curr_story_bad
+            system_action = cls.parse_system_turn(domain_knowledge, line)
+            system_action_name = system_action.get("dialog_acts")[0].get("act")
+
+            for curr_story_utters in curr_story_utters_batch:
+                if cls.last_turn_is_systems_turn(curr_story_utters):
+                    # deal with consecutive system actions by inserting the last user replics in between
+                    curr_story_utters.append(cls.get_last_users_turn(curr_story_utters))
+
+            def parse_form_name(story_line):
```
Comment (oserikov) [inline deeppavlov/dataset_readers/md_yaml_dialogs_reader.py:281]: todo: docs, typehints

File: deeppavlov/dataset_readers/md_yaml_dialogs_reader.py [hunk 4]
```diff
@@ -241,78 +246,96 @@ def _read_story(cls,
...
+                form_name = parse_form_name(system_action_name)
+                augmented_utters = cls.augment_form(form_name, domain_knowledge, intent2slots2text)
+                utters_to_append_batch = [[processed_line for processed_line in processed_line_batch]
+                                                 for aug_line in augmented_utters
+                                                 for processed_line_batch in process_story_line(aug_line)]
+            else:
+                utters_to_append_batch = [[system_action]]
+            return utters_to_append_batch
+
+        def process_story_line(line):
```
Comment (oserikov) [inline deeppavlov/dataset_readers/md_yaml_dialogs_reader.py:298]: todo: docs, typehints

File: deeppavlov/dataset_readers/md_yaml_dialogs_reader.py [hunk 5]
```diff
@@ -333,6 +356,89 @@ def _read_story(cls,
 
         return gobot_formatted_stories
 
+    @classmethod
+    def augment_form(cls, form_name:str, domain_knowledge: DomainKnowledge, intent2slots2text: Dict):
```
Comment (oserikov) [inline deeppavlov/dataset_readers/md_yaml_dialogs_reader.py:360]: todo: docs, returned typehints

File: deeppavlov/dataset_readers/md_yaml_dialogs_reader.py [hunk 6]
```diff
@@ -333,6 +356,89 @@ def _read_story(cls,
...
+        for slot_name, slot_info_li in form.items():
+            if slot_info_li and slot_info_li[0].get("type", '') == "from_entity":
+                # we only handle this sort of slots
+                known_responses = list(domain_knowledge.response_templates)
+                known_intents = list(intent2slots2text.keys())
+                augmended_story.extend(cls.augment_slot(known_responses, known_intents, slot_name, form_name))
+        return augmended_story
+
+    @classmethod
+    def augment_slot(cls, known_responses: List[str], known_intents: List[str], slot_name: str, form_name: str):
```
Comment (oserikov) [inline deeppavlov/dataset_readers/md_yaml_dialogs_reader.py:372]: todo: docs, returned typehints

File: deeppavlov/dataset_readers/md_yaml_dialogs_reader.py [hunk 7]
```diff
@@ -333,6 +356,89 @@ def _read_story(cls,
...
+        ask_slot_act_name = cls.get_augmented_ask_slot_utter(form_name, known_responses, slot_name)
+        inform_slot_user_utter = cls.get_augmented_ask_intent_utter(known_intents, slot_name)
+
+        return [f"- {ask_slot_act_name}", f"* {inform_slot_user_utter}"]
+
+    @classmethod
+    def get_augmented_ask_intent_utter(cls, known_intents, slot_name):
```
Comment (oserikov) [inline deeppavlov/dataset_readers/md_yaml_dialogs_reader.py:379]: todo: docs, returned typehints

File: deeppavlov/dataset_readers/md_yaml_dialogs_reader.py [hunk 8]
```diff
@@ -333,6 +356,89 @@ def _read_story(cls,
...
+        if inform_slot_user_utter_hypothesis in known_intents:
+            inform_slot_user_utter = inform_slot_user_utter_hypothesis
+        else:
+            # todo raise an exception
+            inform_slot_user_utter = None
+            pass
+        return inform_slot_user_utter
+
+    @classmethod
+    def get_augmented_ask_slot_utter(cls, form_name, known_responses, slot_name):
```
Comment (oserikov) [inline deeppavlov/dataset_readers/md_yaml_dialogs_reader.py:390]: todo: docs, returned typehints

File: deeppavlov/dataset_readers/md_yaml_dialogs_reader.py [hunk 9]
```diff
@@ -333,6 +356,89 @@ def _read_story(cls,
...
+        elif ask_slot_act_name_hypothesis2 in known_responses:
+            ask_slot_act_name = ask_slot_act_name_hypothesis2
+        else:
+            # todo raise an exception
+            ask_slot_act_name = None
+            pass
+        return ask_slot_act_name
+
+    @classmethod
+    def get_last_users_turn(cls, curr_story_utters):
```
Comment (oserikov) [inline deeppavlov/dataset_readers/md_yaml_dialogs_reader.py:404]: todo: docs, returned typehints

File: deeppavlov/dataset_readers/md_yaml_dialogs_reader.py [hunk 10]
```diff
@@ -333,6 +356,89 @@ def _read_story(cls,
...
+        *_, last_user_utter = filter(lambda x: x["speaker"] == cls._USER_SPEAKER_ID, curr_story_utters)
+        return last_user_utter
+
+    @classmethod
+    def last_turn_is_systems_turn(cls, curr_story_utters):
```
Comment (oserikov) [inline deeppavlov/dataset_readers/md_yaml_dialogs_reader.py:409]: todo: docs, returned typehints

File: deeppavlov/dataset_readers/md_yaml_dialogs_reader.py [hunk 11]
```diff
@@ -333,6 +356,89 @@ def _read_story(cls,
...
+        return curr_story_utters and curr_story_utters[-1]["speaker"] == cls._SYSTEM_SPEAKER_ID
+
+    @classmethod
+    def parse_system_turn(cls, domain_knowledge, line):
```
Comment (oserikov) [inline deeppavlov/dataset_readers/md_yaml_dialogs_reader.py:413]: todo: docs, returned typehints

File: deeppavlov/dataset_readers/md_yaml_dialogs_reader.py [hunk 12]
```diff
@@ -333,6 +356,89 @@ def _read_story(cls,
...
+        curr_action_text = cls._system_action2text(domain_knowledge, system_action_name)
+        system_action = {"speaker": cls._SYSTEM_SPEAKER_ID,
+                         "text": curr_action_text,
+                         "dialog_acts": [{"act": system_action_name, "slots": []}]}
+        if system_action_name.startswith("action"):
+            system_action["db_result"] = {}
+        return system_action
+
+    @classmethod
+    def augment_user_turn(cls, intent2slots2text, line, slot_name2text2value):
```
Comment (oserikov) [inline deeppavlov/dataset_readers/md_yaml_dialogs_reader.py:425]: todo: docs, returned typehints

File: deeppavlov/models/go_bot/nlg/mock_json_nlg_manager.py [hunk 13]
```diff
@@ -140,3 +142,10 @@ def num_of_known_actions(self) -> int:
             the number of actions known to the NLG module
         """
         return len(self.action_tuples2ids.keys())
+
+    def known_actions(self) -> List:
+        """
+        Returns:
+             the list of actions known to the NLG module
+        """
+        return list(self.action_tuples2ids.keys())
```
Comment (oserikov) [inline deeppavlov/models/go_bot/nlg/mock_json_nlg_manager.py:151]: todo: newline

File: deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py [hunk 14]
```diff
@@ -33,28 +34,76 @@ def get_current_knowledge(self) -> DSTKnowledge:
...
+                                                      parent_tracker.stories_path)
+
+        dialogue_state_tracker.ffill_act_ids2req_slots_ids = action_id2req_slots_ids
+        dialogue_state_tracker.ffill_act_ids2aqd_slots_ids = action_id2aqd_slots_ids
+
+        # endregion set formfilling info
+        return dialogue_state_tracker
+
+    @staticmethod
+    def extract_reqiured_acquired_slots_ids_mapping(act2act_id, slot_names, nlg_manager, parent_tracker):
```
Comment (oserikov) [inline deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py:85]: todo: docs, returned typehints

File: deeppavlov/models/go_bot/tracker/featurized_tracker.py [hunk 15]
```diff
@@ -105,3 +122,124 @@ def _new_features(self, state) -> np.ndarray:
...
+        domain_yml_path = expand_path(domain_yml_path)
+        domain_knowledge: DomainKnowledge = DomainKnowledge.from_yaml(domain_yml_path)
+        potential_api_or_db_actions = domain_knowledge.known_actions
+        forms = domain_knowledge.forms
+        form_names = list(forms.keys())
+
+        # todo migrate to rasa2.0
+        # stories_yml_path = expand_path(stories_yml_path)
+        # stories_yml_di = read_yaml(stories_yml_path)
+        def read_md_story(story_path):
```
Comment (oserikov) [inline deeppavlov/models/go_bot/tracker/featurized_tracker.py:171]: todo: docs, returned typehints

File: deeppavlov/models/go_bot/tracker/featurized_tracker.py [hunk 16]
```diff
@@ -105,3 +122,124 @@ def _new_features(self, state) -> np.ndarray:
...
+                if curr_action.startswith("form"):
+                    curr_action = json.loads(curr_action[len("form"):])["name"]
+                    print(curr_action)
+                if curr_action in form_names:
+                    prev_forms.append(curr_action)
+                if curr_action in potential_api_or_db_actions:
+                    action2forms[curr_action] = prev_forms
+                    prev_forms = []
+
+        def get_slots(system_utter, form_name):
```
Comment (oserikov) [inline deeppavlov/models/go_bot/tracker/featurized_tracker.py:218]: todo: docs, returned typehints

File: deeppavlov/models/go_bot/tracker/featurized_tracker.py [hunk 17]
```diff
@@ -105,3 +122,124 @@ def _new_features(self, state) -> np.ndarray:
...
+                                  MD_YAML_DialogsDatasetReader.augment_form(form_name, domain_knowledge, {})
+                                  if utter.strip().startswith("-")}
+        forms2acquired_slots = {form_name: self._get_form_acquired_slots(form) for form_name, form in forms.items()}
+        actions2required_slots = {act: {slot
+                                        for form in forms
+                                        for slot in forms2acquired_slots[form]}
+                                  for act, forms in action2forms.items()}
+        return actions2required_slots, actions2acquired_slots
+
+    def _get_form_acquired_slots(self, form):
```
Comment (oserikov) [inline deeppavlov/models/go_bot/tracker/featurized_tracker.py:241]: todo: docs, returned typehints

File: deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py [hunk 18]
```diff
@@ -33,28 +34,91 @@ def get_current_knowledge(self) -> DSTKnowledge:
...
+        dialogue_state_tracker.ffill_act_ids2aqd_slots_ids = action_id2aqd_slots_ids
+
+        # endregion set formfilling info
+        return dialogue_state_tracker
+
+    @staticmethod
+    def extract_reqiured_acquired_slots_ids_mapping(act2act_id: Dict,
+                                                    slot_names: List,
+                                                    nlg_manager: NLGManagerInterface,
+                                                    parent_tracker: FeaturizedTracker) -> Tuple[Dict[str, np.ndarray], Dict[str, np.ndarray]]:
```
Comment (IgnatovFedor) [inline deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py:89]:
```suggestion
                                                    parent_tracker: FeaturizedTracker) -> Tuple[Dict[str, np.ndarray],
                                                                                                Dict[str, np.ndarray]]:
```

File: deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py [hunk 19]
```diff
@@ -33,28 +34,91 @@ def get_current_knowledge(self) -> DSTKnowledge:
...
+
+            action_id2req_slots_ids[act_id] = np.zeros(len(slot_names), dtype=np.float32)
+            action_id2aqd_slots_ids[act_id] = np.zeros(len(slot_names), dtype=np.float32)
+
+            if isinstance(act, tuple):
+                acts = act
+            else:
+                acts = [act]
+
+            for act in acts:
```
Comment (IgnatovFedor) [inline deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py:114]: Variable 'act' already declared in 'for' loop statement above. Not a mistake but not a best practice either

File: deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py [hunk 20]
```diff
@@ -111,6 +175,16 @@ def calc_action_mask(self) -> np.ndarray:
             if prev_act_id == self.api_call_id:
                 mask[prev_act_id] = 0.
 
+        for act_id in range(self.n_actions):
+            required_slots_mask = self.ffill_act_ids2req_slots_ids[act_id]
+            acquired_slots_mask = self.ffill_act_ids2aqd_slots_ids[act_id]
+            act_req_slots_fulfilled = np.equal((required_slots_mask * self._binary_features()), required_slots_mask)
+            act_requirements_not_fulfilled = np.invert(act_req_slots_fulfilled)# if act_req_slots_fulfilled != [] else np.array([])
```
Comment (IgnatovFedor) [inline deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py:182]: Is it necessary to keep this comment?

File: deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py [hunk 21]
```diff
@@ -166,6 +240,7 @@ def get_user_tracker(self, user_id: int) -> DialogueStateTracker:
         return tracker
 
     def new_tracker(self):
+        # todo deprecated and never used?
```
Comment (IgnatovFedor) [inline deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py:243]: Please, clarify this comment

File: deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py [hunk 22]
```diff
@@ -185,8 +260,12 @@ def init_new_tracker(self, user_id: int, tracker_entity: DialogueStateTracker) -
             tracker_entity.n_actions,
             tracker_entity.api_call_id,
             tracker_entity.hidden_size,
-            tracker_entity.database
+            tracker_entity.database,
+            tracker_entity.domain_yml_path,
+            tracker_entity.stories_path
         )
+        tracker.ffill_act_ids2req_slots_ids = tracker_entity.ffill_act_ids2req_slots_ids
+        tracker.ffill_act_ids2aqd_slots_ids = tracker_entity.ffill_act_ids2aqd_slots_ids
```
Comment (IgnatovFedor) [inline deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py:268]: Same as comment on lines +79 to +80

File: deeppavlov/models/go_bot/tracker/featurized_tracker.py [hunk 23]
```diff
@@ -1,8 +1,13 @@
-from typing import List, Iterator
+import json
+from pathlib import Path
+from typing import List, Iterator, Union, Optional, Dict, Tuple
 
 import numpy as np
 
+from deeppavlov.core.commands.utils import expand_path
+from deeppavlov.core.common.file import read_yaml
```
Comment (IgnatovFedor) [inline deeppavlov/models/go_bot/tracker/featurized_tracker.py:8]: imported but isn't used

File: deeppavlov/models/go_bot/tracker/featurized_tracker.py [hunk 24]
```diff
@@ -18,14 +23,26 @@ class FeaturizedTracker(TrackerInterface):
...
     """
 
     def get_current_knowledge(self) -> TrackerKnowledgeInterface:
         raise NotImplementedError("Featurized tracker lacks get_current_knowledge() method. "
                                   "To be improved in future versions.")
 
-    def __init__(self, slot_names: List[str]) -> None:
+    def __init__(self,
+                 slot_names: List[str],
+                 # actions_required_acquired_slots_path: Optional[Union[str, Path]]=None,
```
Comment (IgnatovFedor) [inline deeppavlov/models/go_bot/tracker/featurized_tracker.py:36]:
```suggestion
```

File: deeppavlov/models/go_bot/tracker/featurized_tracker.py [hunk 25]
```diff
@@ -18,14 +23,26 @@ class FeaturizedTracker(TrackerInterface):
...
+                 domain_yml_path: Optional[Union[str, Path]]=None,
+                 stories_yml_path: Optional[Union[str, Path]]=None,
+                 **kwargs) -> None:
         self.slot_names = list(slot_names)
+        self.domain_yml_path = domain_yml_path
+        self.stories_path = stories_yml_path
+        self.action_names2required_slots, self.action_names2acquired_slots =\
+            self._load_actions2slots_formfilling_info_from(domain_yml_path, stories_yml_path)
+        # self._load_actions2slots_formfilling_info_from_json(self.actions_required_acquired_slots_path)
```
Comment (IgnatovFedor) [inline deeppavlov/models/go_bot/tracker/featurized_tracker.py:45]:
```suggestion
```

File: deeppavlov/models/go_bot/tracker/featurized_tracker.py [hunk 26]
```diff
@@ -105,3 +122,152 @@ def _new_features(self, state) -> np.ndarray:
...
+
+        domain_yml_path = expand_path(domain_yml_path)
+        domain_knowledge: DomainKnowledge = DomainKnowledge.from_yaml(domain_yml_path)
+        potential_api_or_db_actions = domain_knowledge.known_actions
+        forms = domain_knowledge.forms
+        form_names = list(forms.keys())
+
+        # todo migrate to rasa2.0
+        # stories_yml_path = expand_path(stories_yml_path)
+        # stories_yml_di = read_yaml(stories_yml_path)
```
Comment (IgnatovFedor) [inline deeppavlov/models/go_bot/tracker/featurized_tracker.py:173]:
```suggestion
```

File: deeppavlov/models/go_bot/tracker/featurized_tracker.py [hunk 27]
```diff
@@ -105,3 +122,152 @@ def _new_features(self, state) -> np.ndarray:
...
+            Returns:
+                the slots extracted from the line
+            """
+            slots = []
+            if system_utter.startswith(f"utter_ask_{form_name}_"):
+                slots.append(system_utter[len(f"utter_ask_{form_name}_"):])
+            elif system_utter.startswith(f"utter_ask_"):
+                slots.append(system_utter[len(f"utter_ask_"):])
+            else:
+                # todo: raise an exception
```
Comment (IgnatovFedor) [inline deeppavlov/models/go_bot/tracker/featurized_tracker.py:245]: What prevents you from raise exception now?

File: deeppavlov/dataset_readers/md_yaml_dialogs_reader.py [hunk 28]
```diff
@@ -241,78 +246,142 @@ def _read_story(cls,
...
+                        possible_extensions = process_story_line(user_utter)
+                        for possible_extension in possible_extensions:
+                            new_curr_story_utters = curr_story_utters.copy()
+                            new_curr_story_utters.extend(possible_extension)
+                            new_curr_story_utters_batch.append(new_curr_story_utters)
+                    utters_to_append_batch = new_curr_story_utters_batch
+
+                # utters_to_append_batch = [[processed_line for processed_line in processed_line_batch]
+                #                                  for aug_line in augmented_utters
+                #                                  for processed_line_batch in process_story_line(aug_line)]
```
Comment (IgnatovFedor) [inline deeppavlov/dataset_readers/md_yaml_dialogs_reader.py:329]:
```suggestion
```

File: deeppavlov/dataset_readers/md_yaml_dialogs_reader.py [hunk 29]
```diff
@@ -241,78 +246,142 @@ def _read_story(cls,
...
+                # dialogs MUST start with system replics
+                for curr_story_utters in curr_story_utters_batch:
+                    if not curr_story_utters:
+                        curr_story_utters.append(default_system_start)
+
+                utters_to_append_batch = []
+                for user_utter in possible_user_utters:
+                    utters_to_append_batch.append([user_utter])
+
+            except KeyError as e:
```
Comment (IgnatovFedor) [inline deeppavlov/dataset_readers/md_yaml_dialogs_reader.py:272]:
```suggestion
            except KeyError:
```
you don't use KeyError object

**Review by IgnatovFedor — CHANGES_REQUESTED** [review]

File: deeppavlov/models/go_bot/tracker/featurized_tracker.py (hunk 27 above)
Comment (oserikov) [reply deeppavlov/models/go_bot/tracker/featurized_tracker.py:245]: I don't have time to properly handle it...

**Pull request merged**